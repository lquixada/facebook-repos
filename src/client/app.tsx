import React from 'react'
import {AppContainer} from 'react-hot-loader'
import {Provider} from 'react-redux'
import {renderRoutes} from 'react-router-config'
import {BrowserRouter} from 'react-router-dom'

import routes from '../shared/routes'
import store from './store'

const App: any = AppContainer

export default () => (
  <App>
    <Provider store={store}>
      <BrowserRouter>
        {renderRoutes(routes)}
      </BrowserRouter>
    </Provider>
  </App>
)
