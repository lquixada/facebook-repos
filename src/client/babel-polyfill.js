// Issue and solution: https://github.com/babel/babel-loader/issues/401#issuecomment-365650944
if (!global._babelPolyfill) {
  require('babel-polyfill')
}
