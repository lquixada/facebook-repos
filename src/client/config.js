// Client-side config
const apiHost = process.env.NODE_ENV === 'production' ? 'https://repos.lquixada.com' : 'http://localhost:3001'

export default {
  staticUrl: process.env.NODE_ENV === 'production' ? 'https://static.lquixada.com/repos' : '/assets',
  graphql: `${apiHost}/graphql`,
  github: 'https://github.com'
}
