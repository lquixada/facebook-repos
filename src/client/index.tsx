import React from 'react'
import ReactDOM from 'react-dom'
import {AppContainer} from 'react-hot-loader'
import {Provider} from 'react-redux'
import {renderRoutes} from 'react-router-config'
import {BrowserRouter} from 'react-router-dom'
import 'regenerator-runtime/runtime'

import store from './store'

const App: any = AppContainer
const renderApp = () => {
  const routes = require('../shared/routes').default

  ReactDOM.hydrate(
    <App>
      <Provider store={store}>
        <BrowserRouter>
          {renderRoutes(routes)}
        </BrowserRouter>
      </Provider>
    </App>
    , document.getElementById('app'),
  )
}

renderApp()

if (module.hot) {
  module.hot.accept('../shared/routes', renderApp)
}
