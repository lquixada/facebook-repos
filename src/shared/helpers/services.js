import 'cross-fetch/polyfill'
import gql from 'graphql-tag'

import {getClient} from './client'

export const fetchContributors = async (repoName, page) => {
  const client = getClient()
  const {data} = await client.query({
    query: gql`
      {
        contributors(repo: "${repoName}") {
          nextPage
          data {
            login
            html_url
            avatar_url
          }
        }
      }
    `})

  return data.contributors
}

export const fetchRepo = async (repoName) => {
  const client = getClient()
  const {data} = await client.query({
    query: gql`
      {
        repo(name: "${repoName}") {
          name
          description
          html_url
          stargazers_count
          forks_count
          subscribers_count
          open_issues_count
          license {
            name
          }
        }
      }
    `})

  return data.repo
}

export const fetchRepos = async () => {
  const client = getClient()
  const {data} = await client.query({
    query: gql`
      {
        repoCount {
          name
          count
        }
      }
    `})

  return data.repoCount
    // Convert result
    .map(repo => [repo.name, repo.count])
    // Sort result
    .sort((a, b) => b[1] - a[1])
}
