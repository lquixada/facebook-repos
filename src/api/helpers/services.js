import fetch from 'cross-fetch'

import {extractNext, extractTotal} from '../../shared/helpers'
import config from '../config'

const fetchJson = async (url) => {
  const separator = (url.indexOf('?') > -1 ? '&' : '?')
  const response = await fetch(`${url}${separator}access_token=${config.github.accessToken}`)
  const json = await response.json()

  return {response, json}
}

export const fetchContributorsCount = async (repoName) => {
  const url = config.github.contributorsUrl.replace(':repo', repoName)
  const {response} = await fetchJson(`${url}?page=1&per_page=1`)

  return extractTotal(response.headers.get('Link'))
}

export const fetchContributors = async (url) => {
  const {response, json} = await fetchJson(url)

  return {
    next: extractNext(response.headers.get('Link')),
    result: json
  }
}

export const fetchRepo = async (repoName) => {
  const url = config.github.repoUrl.replace(':repo', repoName)
  return fetchJson(url)
}

export const fetchRepos = async () => {
  const {json} = await fetchJson(config.github.reposUrl)
  return json
}
