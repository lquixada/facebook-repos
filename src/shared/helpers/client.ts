import ApolloClient from 'apollo-boost'
import config from '../../server/config'

const client = new ApolloClient({
  uri: config.graphql,
})

export const getClient = () => client
