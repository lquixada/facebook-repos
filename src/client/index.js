import 'babel-polyfill';
import React from 'react';
import ReactDOM from 'react-dom';
import {Provider} from 'react-redux';
import {BrowserRouter} from 'react-router-dom';
import {renderRoutes, matchRoutes} from 'react-router-config';
import {AppContainer} from 'react-hot-loader';
import {END} from 'redux-saga';

import {trigger} from '../shared/helpers';
import routes from '../shared/routes'; // eslint-disable-line no-unused-vars
import store from './store';

const renderApp = () => {
  const routes = require('../shared/routes').default;

  ReactDOM.hydrate(
    <AppContainer>
      <Provider store={store}>
        <BrowserRouter>
          {renderRoutes(routes)}
        </BrowserRouter>
      </Provider>
    </AppContainer>
    , document.getElementById('app')
  );
};

const matchs = matchRoutes(routes, window.location.pathname);

if (matchs.length === 0) {
  throw new Error(`React Router: Not found ${window.location.pathname}`);
}

if (!window.__INITIAL_STATE__) {
  store.runnedSagas.toPromise().then(renderApp);
  trigger('fetch', matchs, store.dispatch);
  store.dispatch(END);
} else {
  renderApp();
}

if (module.hot) {
  module.hot.accept('../shared/routes', renderApp);
}
