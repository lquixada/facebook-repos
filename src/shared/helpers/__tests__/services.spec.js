import createMockClient from '../../../__tests__/mocked-client'
import {fetchContributors, fetchRepo, fetchRepos} from '../services'

import * as helpers from '../client'

describe('Helpers (Services)', () => {
  beforeEach(() => {
    const client = createMockClient()
    helpers.getClient = () => client
  })

  describe('fetchContributors', () => {
    it('returns contributors from a repo', async () => {
      const data = await fetchContributors('some-repo')
      expect(data).toMatchSnapshot()
    })
  })

  describe('fetchRepo', () => {
    it('returns data from a repo', async () => {
      const data = await fetchRepo('some-repo')
      expect(data).toMatchSnapshot()
    })
  })

  describe('fetchRepos', () => {
    it('returns repos from an organization', async () => {
      const data = await fetchRepos()
      expect(data).toMatchSnapshot()
    })
  })
})
