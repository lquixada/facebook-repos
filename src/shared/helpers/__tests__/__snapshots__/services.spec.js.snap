// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Helpers (Services) fetchContributors returns contributors from a repo 1`] = `
Object {
  "data": Array [
    Object {
      "avatar_url": "http://avatar_url/",
      "html_url": "http://html_url/",
      "login": "user1",
      Symbol(id): "$ROOT_QUERY.contributors({\\"repo\\":\\"some-repo\\"}).data.0",
    },
  ],
  "nextPage": 2,
  Symbol(id): "$ROOT_QUERY.contributors({\\"repo\\":\\"some-repo\\"})",
}
`;

exports[`Helpers (Services) fetchRepo returns data from a repo 1`] = `
Object {
  "description": "",
  "forks_count": 1,
  "html_url": "",
  "license": Object {
    "name": "",
    Symbol(id): "$ROOT_QUERY.repo({\\"name\\":\\"some-repo\\"}).license",
  },
  "name": "repo1",
  "open_issues_count": 1,
  "stargazers_count": 1,
  "subscribers_count": 1,
  Symbol(id): "$ROOT_QUERY.repo({\\"name\\":\\"some-repo\\"})",
}
`;

exports[`Helpers (Services) fetchRepos returns repos from an organization 1`] = `
Array [
  Array [
    "repo1",
    2,
  ],
]
`;
